### PR TITLE
fix: detect external autoscaler policy changes

### PR DESCRIPTION
**Pull request by laimonasr**

### CODE REVIEW

File: castai/resource_autoscaler.go [hunk 1]
```diff
@@ -5,12 +5,12 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	jsonpatch "github.com/evanphx/json-patch"
 	"io"
 	"log"
 	"net/http"
 	"time"
 
-	jsonpatch "github.com/evanphx/json-patch"
```
Comment (viktorasm) [inline castai/resource_autoscaler.go:13]: nitpick: this moved to the wrong group for no reason

File: castai/utils.go [hunk 2]
```diff
@@ -107,3 +111,12 @@ func toNilList[T any](l *[]T) *[]T {
 	}
 	return l
 }
+
+func normalizeJSON(bytes []byte) ([]byte, error) {
```
Comment (viktorasm) [inline castai/utils.go:115]: what is this supposed to achieve?

File: castai/resource_autoscaler.go [hunk 3]
```diff
@@ -74,6 +75,23 @@ func resourceCastaiAutoscalerDelete(ctx context.Context, data *schema.ResourceDa
...
+	clusterId := getClusterId(d)
+	if clusterId == "" {
+		return nil
+	}
+
+	policies, err := getChangedPolicies(ctx, d, meta, clusterId)
+	if err != nil {
+		return err
+	}
+	if policies == nil {
```
Comment (viktorasm) [inline castai/resource_autoscaler.go:88]: probably a redundant check as method does not return nil,nil

Comment (viktorasm) [reply castai/resource_autoscaler.go:88]: ah, nevermind, I can see one path has a nil,nil result.

**Review by viktorasm — APPROVED** [review]

File: castai/utils.go (hunk 2 above)
Comment (laimonasr) [reply castai/utils.go:115]: Umarshal and marshal produces the same json string when the input json has differently ordered keys. Terraform would detect whitespace changes when json values are the same but their keys are ordered differently so this function normalizes the key order

File: castai/resource_autoscaler.go (hunk 1 above)
Comment (laimonasr) [reply castai/resource_autoscaler.go]: Moved it back

File: castai/utils.go (hunk 2 above)
Comment (viktorasm) [reply castai/utils.go:115]: can you stress test it with a sample JSON? I think it uses maps as an intermediate holding structure, and those do not always result in same iteration order, e.g. tests based on map key order tend to be flaky for this reason.

Comment (laimonasr) [reply castai/utils.go:115]: Added a unit test with a few test cases

**Review by aldor007 — APPROVED** [review]

**Pull request merged**